cirq/devices/grid_qubit.py

```diff
@@ -1,157 +1 @@
                empty locations in the grid. If diagram has characters other
                than alphanumerics, spacers, and newlines ('\n'), an error will
                be thrown. The top-left corner of the diagram will be have
                coordinate (0,0).

        Returns:
            A list of GridQubits corresponding to the provided diagram

        Raises:
            ValueError: If the input string contains an invalid character.
        """
        lines = diagram.strip().split('\n')
        no_qubit_characters = ['.', '-', ' ']
        qubits = []
        for row, line in enumerate(lines):
            for col, c in enumerate(line.strip()):
                if c not in no_qubit_characters:
                    if not c.isalnum():
                        raise ValueError("Input string has invalid character")
                    qubits.append(GridQubit(row, col))
        return qubits

    def __repr__(self):
        return 'cirq.GridQubit({}, {})'.format(self.row, self.col)

    def __str__(self):
        return '({}, {})'.format(self.row, self.col)

    def __add__(self, other: Tuple[int, int]) -> 'GridQubit':
        assert isinstance(other, tuple) and len(other) == 2 and all(
            isinstance(x, int) for x in other), (
            'Can only add tuple of length 2 to GridQubit. Was {}'.format(other))
        return GridQubit(row=self.row + other[0], col=self.col + other[1])

    def __sub__(self, other: Tuple[int, int]) -> 'GridQubit':
        assert isinstance(other, tuple) and len(other) == 2 and all(
            isinstance(x, int) for x in other), (
            'Can only add tuple of length 2 to GridQubit. Was {}'.format(other))
        return GridQubit(row=self.row - other[0], col=self.col - other[1])

    def __radd__(self, other: Tuple[int, int]) -> 'GridQubit':
        return self + other

    def __rsub__(self, other: Tuple[int, int]) -> 'GridQubit':
        return - self + other

    def __neg__(self) -> 'GridQubit':
        return GridQubit(row=-self.row, col=-self.col)

    def to_proto_dict(self, v2_proto=False) -> Dict:
        """Return the proto in dictionary form."""
        # TODO: Deprecate v1 proto method.
        return {
            'row': self.row,
            'col': self.col,
        }

    def proto_id(self) -> str:
        return '{}_{}'.format(self.row, self.col)

    @staticmethod
    def from_proto_dict(proto_dict: Dict) -> 'GridQubit':
        """Proto dict must have 'row' and 'col' keys."""
        # TODO: Deprecate v1 proto method.
        if 'row' not in proto_dict or 'col' not in proto_dict:
            raise ValueError(
                'Proto dict does not contain row or col: {}'.format(proto_dict))
        return GridQubit(row=proto_dict['row'], col=proto_dict['col'])

    @staticmethod
    def from_proto_id(proto_id: str) -> 'GridQubit':
        row, col = proto_id.split('_')
        return GridQubit(row=int(row), col=int(col))

```

Comment:
Your editor mangled the file.

Reply:
The formater isn't even upset about it! Very strange.

Reply:
wowah